epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala

```diff
@@ -0,0 +1,300 @@
+        val counter = readReadyCounter + 1
+        readReadyCounter = counter
+        val cb = readCallback
+        readCallback = null
+        if (cb ne null) cb(Right(counter))
+      }
+      if ((events & EPOLLOUT) != 0) {
+        val counter = writeReadyCounter + 1
+        writeReadyCounter = counter
+        val cb = writeCallback
+        writeCallback = null
+        if (cb ne null) cb(Right(counter))
+      }
+    }
+
+    def pollReadRec[A, B](a: A)(f: A => IO[Either[A, B]]): IO[B] =
+      readMutex.lock.surround {
+        def go(a: A, before: Int): IO[B] =
+          // invoke `f`
+          f(a).flatMap {
+            case Left(a) =>
+              // `f` fails because the file descriptor is blocked
+              IO(readReadyCounter).flatMap { after =>
+                if (before != after)
+                  // there was a read-ready notification since we started, try again immediately
+                  go(a, after)
+                else
+                  IO.asyncCheckAttempt[Int] { cb =>
+                    IO {
+                      readCallback = cb
+                      // check again before we suspend
+                      val now = readReadyCounter
+                      if (now != before) {
+                        readCallback = null
+                        Right(now)
+                      } else {
+                        // block here until `readCallback` is invoked
+                        Left(Some(IO(this.readCallback = null)))
+                      }
+                    }
+                  }.flatMap(go(a, _))
+              }
+            case Right(b) => IO.pure(b)
+          }
+
+        IO(readReadyCounter).flatMap(go(a, _))
+      }
+
+    def pollWriteRec[A, B](a: A)(f: A => IO[Either[A, B]]): IO[B] =
+      writeMutex.lock.surround {
+        def go(a: A, before: Int): IO[B] =
+          f(a).flatMap {
+            case Left(a) =>
+              IO(writeReadyCounter).flatMap { after =>
+                if (before != after)
+                  // there was a write-ready notification since we started, try again immediately
+                  go(a, after)
+                else
+                  IO.asyncCheckAttempt[Int] { cb =>
+                    IO {
+                      writeCallback = cb
+                      // check again before we suspend
+                      val now = writeReadyCounter
+                      if (now != before) {
+                        writeCallback = null
+                        Right(now)
+                      } else Left(Some(IO(this.writeCallback = null)))
+                    }
+                  }.flatMap(go(a, _))
+              }
+            case Right(b) => IO.pure(b)
+          }
+
+        IO(writeReadyCounter).flatMap(go(a, _))
+      }
+  }
+
+  final class Poller private[EpollSystem] (epfd: Int, evfd: Int) {
+
+    // fixed array used only when reading from eventfd
+    private final lazy val readBuf = Memory.allocate(globalRuntime, 8)
```

Comment:
We should avoid `lazy val` unless there is a strong motivation reason why it should be lazy. `lazy val`s add a memory synchronization penalty on every access.

```suggestion
    private[this] val readBuf = Memory.allocate(globalRuntime, 8)
```